src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -505,56 +524,54 @@ private void onMqttSubscribe(
                     }
                 }
 
-                for (int progress = offset; progress < limit; progress = subscription.limit())
+                for (int progress = offset, ackIndex = 0; progress < limit; progress = subscription.limit(), ackIndex++)
                 {
                     subscription = mqttSubscriptionRO.tryWrap(buffer, progress, limit);
                     if (subscription == null)
                     {
                         break;
                     }
-
-                    int reasonCode = 0x8f;
                     final String topicFilter = subscription.topicFilter().asString();
                     final RouteFW route = resolveTarget(routeId, authorization, topicFilter);
 
                     if (route != null)
                     {
-                        final MqttServerStream serverStream = new MqttServerStream();
-
                         final long newRouteId = route.correlationId();
                         final long newInitialId = supplyInitialId.applyAsLong(newRouteId);
                         final long newReplyId = supplyReplyId.applyAsLong(newInitialId);
 
                         final MessageConsumer newTarget = router.supplyReceiver(newInitialId);
 
-                        final MqttBeginExFW beginEx = mqttBeginExRW
-                            .wrap(writeBuffer, BeginFW.FIELD_OFFSET_EXTENSION, writeBuffer.capacity())
-                            .typeId(mqttTypeId)
-                            .role(r -> r.set(MqttRole.RECEIVER))
-                            .clientId("client")
-                            .topic(topicFilter)
-                            .subscriptionId(subscriptionId)
-                            .build();
+                        final MqttServerStream serverStream = new MqttServerStream(this, newTarget,
```

Comment:
Consider moving `MqttServerStream` to an inner class of `MqttServer`, no need to explicitly pass `this`.